core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -4121,6 +4121,102 @@ private void rewriteCountDistinct(QueryModel model) throws SqlException {
         }
     }
 
+    /**
+     * Rewrites "select distinct a,b,c from tab" into "select a,b,c from (select a,b,c,d,count() from tab)".
+     * We will not rewrite distinct of group-by or windows functions.
+     *
+     * @param model input model, it will be validated for the rewrite possibility. If rewrite is impossible,
+     *              the input model will be returned verbatim.
+     * @return either the input model or newly rewritten model.
+     */
+    private QueryModel rewriteDistinct(QueryModel model) throws SqlException {
```

Comment:
The rewrite seems to kick in almost all distinct queries, even where it won't give an improvement, e.g.
```sql
select distinct * from x x1 join x x2 on (c);
select distinct * from (x union x);
```

Is that something expected? Usually we try to narrow the rewrites for the "fast path" case, so that's why I'm asking.

Reply:
this is not the intention as long as we don't make things worse. It seems that group-by produces the same result and having group-by cover more of the executions is advantageous long term. We can pipeline parallel execution in the future. I would leave this in place.

Reply:
I'm a bit worried that we could overlook a combination where things break due to changed order, but the only thing I found is distinct + sample by combination where we keep order by as the top operation - not a big deal.